File: android/app/src/main/java/com/intootv/ExoPlayer.java
package com.intootv;

import android.app.Activity;
import android.app.AlertDialog;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Toast;
import android.view.Window;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.source.hls.HlsMediaSource;
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.MappingTrackSelector;
import com.google.android.exoplayer2.trackselection.MappingTrackSelector.MappedTrackInfo;
import com.google.android.exoplayer2.trackselection.TrackSelection;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
import com.google.android.exoplayer2.ui.DefaultTimeBar;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.ui.TrackSelectionView;
import com.google.android.exoplayer2.upstream.BandwidthMeter;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.upstream.TransferListener;
import com.google.android.exoplayer2.util.Util;
import android.os.Handler;

public class ExoPlayer extends Activity implements View.OnClickListener {

    private static final String KEY_PLAY_WHEN_READY = "play_when_ready";
    private static final String KEY_WINDOW = "window";
    private static final String KEY_POSITION = "position";

    private PlayerView playerView;
    private SimpleExoPlayer player;

    private Timeline.Window window;
    private DataSource.Factory mediaDataSourceFactory;
    private DefaultTrackSelector trackSelector;
    private TrackGroupArray lastSeenTrackGroupArray;
    private boolean shouldAutoPlay;
    private BandwidthMeter bandwidthMeter;

    private ProgressBar progressBar;
    private ImageView ivHideControllerButton;
    private ImageView ivSettings;
    private boolean playWhenReady;
    private int currentWindow;
    private long playbackPosition;
    private String url;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.requestWindowFeature(Window.FEATURE_NO_TITLE); //Remove title bar
        setContentView(R.layout.activity_exo_player);
        this.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN); //Remove notification bar

        if (savedInstanceState == null) {
            Bundle extras = getIntent().getExtras();
            url = extras.getString("url");
            playWhenReady = true;
            currentWindow = 0;
            playbackPosition = 0;
        } else {
            playWhenReady = savedInstanceState.getBoolean(KEY_PLAY_WHEN_READY);
            currentWindow = savedInstanceState.getInt(KEY_WINDOW);
            playbackPosition = savedInstanceState.getLong(KEY_POSITION);
            url = savedInstanceState.getString("url");
        }



        shouldAutoPlay = true;
        bandwidthMeter = new DefaultBandwidthMeter();
        mediaDataSourceFactory = new DefaultDataSourceFactory(this, Util.getUserAgent(this, "mediaPlayerSample"), (TransferListener<? super DataSource>) bandwidthMeter);
        window = new Timeline.Window();
        //  ivHideControllerButton = findViewById(R.id.exo_controller);
        //  ivSettings = findViewById(R.id.settings);
        progressBar = findViewById(R.id.progress_bar);
    }

    private void initializePlayer() {

        playerView = findViewById(R.id.player_view);
        playerView.requestFocus();

        TrackSelection.Factory videoTrackSelectionFactory =
                new AdaptiveTrackSelection.Factory(bandwidthMeter);

        trackSelector = new DefaultTrackSelector(videoTrackSelectionFactory);
        lastSeenTrackGroupArray = null;

        player = ExoPlayerFactory.newSimpleInstance(this, trackSelector);

        playerView.setPlayer(player);

        player.addListener(new PlayerEventListener());
        player.setPlayWhenReady(shouldAutoPlay);

        // Use Hls, Dash or other smooth streaming media source if you want to test the track quality selection.
        Handler mainHandler = new Handler();
        MediaSource mediaSource = new HlsMediaSource(Uri.parse(url),
                mediaDataSourceFactory, mainHandler, null);


//        MediaSource mediaSource = new ExtractorMediaSource.Factory(mediaDataSourceFactory)
//                .createMediaSource(Uri.parse(url));

        boolean haveStartPosition = currentWindow != C.INDEX_UNSET;
        if (haveStartPosition) {
            player.seekTo(currentWindow, playbackPosition);
        }

        player.prepare(mediaSource, !haveStartPosition, false);
        updateButtonVisibilities();

//        ivHideControllerButton.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                playerView.hideController();
//            }
//        });
    }

    private void releasePlayer() {
        if (player != null) {
            updateStartPosition();
            shouldAutoPlay = player.getPlayWhenReady();
            player.release();
            player = null;
            trackSelector = null;
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        if (Util.SDK_INT > 23) {
            initializePlayer();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if ((Util.SDK_INT <= 23 || player == null)) {
            initializePlayer();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (Util.SDK_INT <= 23) {
            releasePlayer();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        if (Util.SDK_INT > 23) {
            releasePlayer();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        updateStartPosition();

        outState.putBoolean(KEY_PLAY_WHEN_READY, playWhenReady);
        outState.putInt(KEY_WINDOW, currentWindow);
        outState.putLong(KEY_POSITION, playbackPosition);
        outState.putString("url", url);
        super.onSaveInstanceState(outState);
    }

    private void updateStartPosition() {
        playbackPosition = player.getCurrentPosition();
        currentWindow = player.getCurrentWindowIndex();
        playWhenReady = player.getPlayWhenReady();
    }

    private void updateButtonVisibilities() {
        //  ivSettings.setVisibility(View.GONE);
        if (player == null) {
            return;
        }

        MappedTrackInfo mappedTrackInfo = trackSelector.getCurrentMappedTrackInfo();
        if (mappedTrackInfo == null) {
            return;
        }

        for (int i = 0; i < mappedTrackInfo.getRendererCount(); i++) {
            TrackGroupArray trackGroups = mappedTrackInfo.getTrackGroups(i);
            if (trackGroups.length != 0) {
                if (player.getRendererType(i) == C.TRACK_TYPE_VIDEO) {
                    // ivSettings.setVisibility(View.VISIBLE);
                    // ivSettings.setOnClickListener(this);
                    // ivSettings.setTag(i);
                }
            }
        }
    }

    @Override
    public void onClick(View v) {
//        if (v.getId() == R.id.settings) {
//            MappedTrackInfo mappedTrackInfo = trackSelector.getCurrentMappedTrackInfo();
//            if (mappedTrackInfo != null) {
//                CharSequence title = getString(R.string.video);
//                int rendererIndex = (int) ivSettings.getTag();
//                Pair<AlertDialog, TrackSelectionView> dialogPair =
//                        TrackSelectionView.getDialog(this, title, trackSelector, rendererIndex);
//                dialogPair.second.setShowDisableOption(false);
//                dialogPair.second.setAllowAdaptiveSelections(true);
//                dialogPair.first.show();
//            }
//        }
    }

    private class PlayerEventListener extends Player.DefaultEventListener{

        @Override
        public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
            switch (playbackState) {
                case Player.STATE_IDLE:       // The player does not have any media to play yet.
                    progressBar.setVisibility(View.VISIBLE);
                    break;
                case Player.STATE_BUFFERING:  // The player is buffering (loading the content)
                    progressBar.setVisibility(View.VISIBLE);
                    break;
                case Player.STATE_READY:      // The player is able to immediately play
                    progressBar.setVisibility(View.GONE);
                    break;
                case Player.STATE_ENDED:      // The player has finished playing the media
                    Log.e("video finished","finish");
                    VideoClassListener.getInstance().videoState(true);
                    progressBar.setVisibility(View.GONE);
                    break;
            }
            updateButtonVisibilities();
        }

        @Override
        public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {
            updateButtonVisibilities();
            // The video tracks are no supported in this device.
            if (trackGroups != lastSeenTrackGroupArray) {
                MappedTrackInfo mappedTrackInfo = trackSelector.getCurrentMappedTrackInfo();
                if (mappedTrackInfo != null) {
                    if (mappedTrackInfo.getTypeSupport(C.TRACK_TYPE_VIDEO)
                            == MappedTrackInfo.RENDERER_SUPPORT_UNSUPPORTED_TRACKS) {
                        Toast.makeText(ExoPlayer.this, "Error unsupported track", Toast.LENGTH_SHORT).show();
                    }
                }
                lastSeenTrackGroupArray = trackGroups;
            }
        }
    }
}

File: android/app/src/main/java/com/intootv/ExoPlayer.kt
package com.intootv

import android.app.Activity
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.WindowManager
import android.widget.ProgressBar
import android.view.Window
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.trackselection.MappingTrackSelector.MappedTrackInfo
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.*
import com.google.android.exoplayer2.util.Util
import com.henninghall.date_picker.DatePickerManager.context
import org.theta.deliverysdk.ThetaDelivery
import org.theta.deliverysdk.datasource.ThetaDataSourceListener
import org.theta.deliverysdk.datasource.ThetaHlsDataSourceFactory
import org.theta.deliverysdk.models.*


private const val KEY_PLAY_WHEN_READY = "play_when_ready"
private const val KEY_WINDOW = "window"
private const val KEY_POSITION = "position"

class ExoPlayer : Activity() {

    private var playerView: PlayerView? = null
    private var player: SimpleExoPlayer? = null
    private var window: Timeline.Window? = null
    private var trackSelector: DefaultTrackSelector? = null
    private var lastSeenTrackGroupArray: TrackGroupArray? = null
    private var shouldAutoPlay = false
    private var progressBar: ProgressBar? = null
    private var playWhenReady = false
    private var currentWindow = 0
    private var playbackPosition: Long = 0
    private val streamUrl = "https://bitdash-a.akamaihd.net/content/MI201109210084_1/m3u8s/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.m3u8";
    private var playerListener: Player.EventListener? = null
    private var bandwidthMeter: BandwidthMeter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE) //Remove title bar
        setContentView(R.layout.activity_exo_player)
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN) //Remove notification bar
        if (savedInstanceState == null) {
            val extras = intent.extras
            playWhenReady = true
            currentWindow = 0
            playbackPosition = 0
        } else {
            playWhenReady = savedInstanceState.getBoolean(KEY_PLAY_WHEN_READY)
            currentWindow = savedInstanceState.getInt(KEY_WINDOW)
            playbackPosition = savedInstanceState.getLong(KEY_POSITION)
        }
        shouldAutoPlay = true
        bandwidthMeter = DefaultBandwidthMeter()
        window = Timeline.Window()
        progressBar = findViewById(R.id.progress_bar)
        ThetaDelivery.init(this);
    }

    private fun initializePlayer() {
        if (player == null) {
            player = ExoPlayerFactory.newSimpleInstance(
                    DefaultRenderersFactory(this),
                    DefaultTrackSelector(),
                    DefaultLoadControl()
            )

            playerView?.player = player
            player?.playWhenReady = playWhenReady
            player?.seekTo(0)
            player?.addListener(getPlayerListener())
        }

        val mediaSource = buildMediaSource(Uri.parse(streamUrl))
        player?.prepare(mediaSource, true, false)
    }
    private fun buildMediaSource(uri: Uri): MediaSource {
        val dataSourceFactory = ThetaHlsDataSourceFactory(
                this,
                Util.getUserAgent(this, "DeliverySDK"),
                BANDWIDTH_METER,
                ThetaConfig(
                        streamUrl,
                        "123",
                        "",
                        "vidwmir59zac1566re4"),
                object : ThetaDataSourceListener {
                    override fun onInfoEvent(thetaInfoEvent: ThetaInfoEvent) {
                        Log.d("PlayerActivity", thetaInfoEvent.message)
                    }

                    override fun onTrafficEvent(trafficEvent: ThetaTrafficEvent) {
                        runOnUiThread {
                        }

                    }

                    override fun onPeersChangedEvent(peersEvent: ThetaPeersChangedEvent) {
                        runOnUiThread {
                        }

                    }

                    override fun onAccountUpdatedEvent(userWalletEvent: ThetaUserWalletEvent) {

                    }

                    override fun onErrorEvent(errorEvent: ThetaErrorEvent) {
                    }
                }
        )

        return HlsMediaSource.Factory(dataSourceFactory).createMediaSource(uri)
    }
    private fun releasePlayer() {
        if (player != null) {
            playWhenReady = player?.playWhenReady ?: true
            player?.release()
            player = null
        }
    }

    public override fun onStart() {
        super.onStart()
        initializePlayer()
    }

    public override fun onResume() {
        super.onResume()
        if (player == null) {
            initializePlayer()
        }
    }

    public override fun onPause() {
        super.onPause()
        releasePlayer()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        updateStartPosition()
        outState.putBoolean(KEY_PLAY_WHEN_READY, playWhenReady)
        outState.putInt(KEY_WINDOW, currentWindow)
        outState.putLong(KEY_POSITION, playbackPosition)
        outState.putString("url", streamUrl)
        super.onSaveInstanceState(outState)
    }

    private fun updateStartPosition() {
        playbackPosition = player!!.currentPosition
        currentWindow = player!!.currentWindowIndex
        playWhenReady = player!!.playWhenReady
    }

    override fun onDestroy() {
        ThetaDelivery.destroy(this)
        super.onDestroy()
    }

    private fun getPlayerListener(): Player.EventListener {
        if (playerListener == null) {
            playerListener = object : Player.EventListener {
                override fun onTimelineChanged(timeline: Timeline, manifest: Any, reason: Int) {
                    if (player != null && player?.bufferedPosition ?: 0 <= -2000) {
                        player?.seekToDefaultPosition()
                    }
                }

                override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {

                }

                override fun onLoadingChanged(isLoading: Boolean) {

                }

                override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {

                }

                override fun onRepeatModeChanged(repeatMode: Int) {

                }

                override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {

                }

                override fun onPlayerError(error: ExoPlaybackException) {
                    Log.d("OnPlayerError", "message: " + error.message)
                }

                override fun onPositionDiscontinuity(reason: Int) {

                }

                override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {

                }

                override fun onSeekProcessed() {

                }
            }
        }

        return playerListener!!
    }


    companion object {

        private val BANDWIDTH_METER = DefaultBandwidthMeter()
    }
}